### PR TITLE
feat: atomic service publish on startup

### PR DESCRIPTION
**Pull request by 3Hren**
This commit changes the core bootstrap mechanism by making its service
publishing atomic. That means that there will be no service
announce/exposing before all of them are ready, i.e. completed their
constructors without exceptions. Doing so protects us from the case,
when an application, which requires some service (which is not exposed
yet) has been started and tries to resolve that service, which results
in errors.

Now there will be no user load to the node before it is completely
bootstrapped.

Also fixed a race condition when an "expose" event on the same service
can be delivered after the "remove" one.

### CODE REVIEW

**Comment by 3Hren** [comment]
@antmat @karitra PTAL.

**Comment by karitra** [comment]
LGTM

File: src/context.cpp [hunk 1]
```diff
@@ -234,33 +242,48 @@ class context_impl_t : public context_t {
...
+            service->prototype()->version(),
+            service->prototype()->root()
         ));
+
+        services.emplace_back(std::move(name), std::move(service));
+    }
+
+    auto
+    publish_all(service_list_t& services) -> void {
+        for (auto it = std::begin(m_unpublished); it != std::end(m_unpublished);) {
```
Comment (antmat) [inline src/context.cpp:281]: Not an issue, but std::begin looks excess, as far it is not generic algo

File: src/context.cpp [hunk 2]
```diff
@@ -234,33 +242,48 @@ class context_impl_t : public context_t {
     insert_with(const std::string& name, std::function<std::unique_ptr<tcp_actor_t>()> fn) override {
```
Comment (antmat) [inline src/context.cpp:242]: Just a note: In general I think 'insert' should only insert service to context without implicit service start. In this case we can get rid of m_bootstrapped flag, and make interface slightly clearer, but it also breaks behaviour.

File: src/context.cpp [hunk 3]
```diff
@@ -161,15 +164,20 @@ class context_impl_t : public context_t {
             }
         });
 
-        if (!errored.empty()) {
+        if (errored.empty()) {
+            m_services.apply([&](service_list_t& list) {
+                publish_all(list);
```
Comment (antmat) [inline src/context.cpp:169]: Can 'run' throw? Previously it was caught during insert.

**Review by antmat — APPROVED** [review]
LGTM in general

**Pull request merged**